test/gurka/test_route_incidents.cc

```diff
@@ -0,0 +1,685 @@
+        << " begins on wrong edge";
+    const auto& end_edge = leg.node(i.end_edge_index).edge();
+    ASSERT_TRUE(end_edge.begin_shape_index() <= incident->end_shape_index() &&
+                incident->end_shape_index() <= end_edge.end_shape_index())
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.end_edge_index
+        << " ends on wrong edge";
+
+    // check that we are on the right shape points
+    const auto& lengths = leg_lengths[i.leg_index];
+    auto begin_edge_length =
+        lengths[begin_edge.end_shape_index()] - lengths[begin_edge.begin_shape_index()];
+    auto begin_edge_pct =
+        (lengths[incident->begin_shape_index()] - lengths[begin_edge.begin_shape_index()]) /
+        begin_edge_length;
+    ASSERT_NEAR(begin_edge_pct, i.begin_pct, .01)
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.begin_edge_index
+        << " begins at wrong location along the edge";
+    auto end_edge_length =
+        lengths[end_edge.end_shape_index()] - lengths[end_edge.begin_shape_index()];
+    auto end_edge_pct =
+        (lengths[incident->end_shape_index()] - lengths[end_edge.begin_shape_index()]) /
+        end_edge_length;
+    ASSERT_NEAR(end_edge_pct, i.end_pct, .1)
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.end_edge_index
+        << " ends at wrong location along the edge";
+  }
+
+  ASSERT_EQ(incident_count, locations.size()) << "Expected number of incidents does not match actual";
+}
+
+// provides us an easy way to mock having incident tiles, each test can override the tile in question
+// a bit more work is needed if we want to do it for more than one tile at a time
+struct test_reader : public baldr::GraphReader {
+  test_reader(const boost::property_tree::ptree& pt) : baldr::GraphReader(pt) {
+    tile_extract_.reset(new baldr::GraphReader::tile_extract_t(pt));
+  }
+  virtual std::shared_ptr<baldr::GraphReader::incident_tile_t>
+  GetIncidentTile(const baldr::GraphId& tile_id) const {
+    return incident_tile_;
+  }
+  void sort() {
+    std::sort(incident_tile_->begin(), incident_tile_->end(),
+              [](const GraphReader::Incident& a, const GraphReader::Incident& b) {
+                if (a.edge_index == b.edge_index) {
+                  if (a.start_offset == b.start_offset) {
+                    if (a.end_offset == b.end_offset) {
+                      return a.id < b.id;
+                    }
+                    return a.end_offset < b.end_offset;
+                  }
+                  return a.start_offset < b.start_offset;
+                }
+                return a.edge_index < b.edge_index;
+              });
+  }
+  std::shared_ptr<baldr::GraphReader::incident_tile_t> incident_tile_;
+};
+
+// used to make a graphreader and mark some edges as having incidents
+std::shared_ptr<test_reader> setup_test(const gurka::map& map,
```

Comment:
this is boilerplate for each test case. it sets up the testable reader and makes the edges that you requested have traffic that says they have incidents so that the graphreader will try to look up their incidents. it returns the reader and the ids of the edges so that the test can use the ids to modify the readers incident data accordingly